hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/listener/ClientNonSmartListenerService.java

```diff
@@ -146,37 +145,27 @@ public void connectionAdded(final Connection connection) {
         registrationExecutor.submit(new Runnable() {
             @Override
             public void run() {
-                Map<ClientRegistrationKey, ClientEventRegistration> tempMap =
-                        new HashMap<ClientRegistrationKey, ClientEventRegistration>();
                 for (ClientRegistrationKey registrationKey : userRegistrations) {
                     try {
                         ClientEventRegistration eventRegistration = invoke(registrationKey);
-                        tempMap.put(registrationKey, eventRegistration);
+                        ClientEventRegistration oldRegistration = activeRegistrations.put(registrationKey, eventRegistration);
+                        if (oldRegistration != null) {
+                            //if there was a registration corresponding to same user listener key before,
+                            //then we need to remove its event handler as cleanup
+                            removeEventHandler(oldRegistration.getCallId());
+                        }
                     } catch (Exception e) {
                         logger.warning("Listener " + registrationKey + " can not be added to new connection: "
```

Comment:
should you remove the event handler regardles of the invoke failures?

Reply:
I am moving the cleanup of event handler before the invocation to cover both success and failure cases